src/Microsoft.ML.Data/Transforms/TermTransformImpl.cs

```diff
@@ -454,8 +451,26 @@ private enum MapType : byte
         /// These are the immutable and serializable analogs to the <see cref="Builder"/> used in
         /// training.
         /// </summary>
-        private abstract class TermMap
+        internal abstract class TermMap
         {
+            private static volatile MemoryStreamPool _codecFactoryPool;
+            private static volatile CodecFactory _codecFactory;
+            private static object _factoryLock = new object();
+
+            private static CodecFactory CodecFactory(IHostEnvironment env)
+            {
+                lock (_factoryLock)
+                {
+                    if (_codecFactory == null)
+                    {
+                        Interlocked.CompareExchange(ref _codecFactoryPool, new MemoryStreamPool(), null);
+                        Interlocked.CompareExchange(ref _codecFactory, new CodecFactory(env, _codecFactoryPool), null);
+                    }
+                }
+                env.Assert(_codecFactory != null);
```

Comment:
Having a lock, *or* using `InterlockedCompare.Exchange`, are both possible choices. (We tend to favor `InterlockedCompare.Exchange` throughout the rest of the codebase, but whatever.) Using *both* however seems somewhat pointless.